internal/controller/atlasschema_controller.go

```diff
@@ -511,6 +549,20 @@ func (r *AtlasSchemaReconciler) extractData(ctx context.Context, res *dbv1alpha1
 			TxMode:  s.TxMode,
 		}
 	)
+	data.Config, err = s.GetConfig(ctx, r, res.Namespace)
+	if err != nil {
+		return nil, transient(err)
+	}
+	if !r.allowCustomConfig && data.Config != nil {
+		return nil, errors.New("install the operator with \"--set allowCustomConfig=true\" to use custom atlas.hcl config")
+	}
+	hasConfig := data.Config != nil
+	if hasConfig && data.EnvName == "" {
+		return nil, errors.New("env name must be set when using custom atlas.hcl config")
+	}
+	if s := s.EnvName; s != "" {
+		data.EnvName = s
+	}
```

Comment:
The check for `EnvName` should check on `s.EnvName`, not `data.EnvName` which always is `defaultEnvName`.

```suggestion
	hasConfig := data.Config != nil
	if hasConfig {
		if !r.allowCustomConfig{
			return nil, errors.New("install the operator with \"--set allowCustomConfig=true\" to use custom atlas.hcl config")
		}
		if s.EnvName == "" {
			return nil, errors.New("env name must be set when using custom atlas.hcl config")
		} else {
			data.EnvName = s
		}
	}
```